lib/apiservers/engine/backends/archive.go

```diff
@@ -0,0 +1,562 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backends
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"encoding/gob"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/tchap/go-patricia/patricia"
+
+	"github.com/vmware/vic/lib/apiservers/engine/backends/cache"
+	viccontainer "github.com/vmware/vic/lib/apiservers/engine/backends/container"
+	vicarchive "github.com/vmware/vic/pkg/archive"
+	"github.com/vmware/vic/pkg/trace"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/pkg/archive"
+)
+
+// ContainerArchivePath creates an archive of the filesystem resource at the
+// specified path in the container identified by the given name. Returns a
+// tar archive of the resource and whether it was a directory or a single file.
+func (c *Container) ContainerArchivePath(name string, path string) (content io.ReadCloser, stat *types.ContainerPathStat, err error) {
+	defer trace.End(trace.Begin(name))
+
+	vc := cache.ContainerCache().GetContainer(name)
+	if vc == nil {
+		return nil, nil, NotFoundError(name)
+	}
+
+	var reader io.ReadCloser
+	reader, err = c.offlineContainerArchivePath(vc, path)
```

Comment:
In the meeting we decided that the online/offline choice would be determined by a try-it-and-see for offline:
1. try the offline mount - vSphere will reject it if the disk is locked
2. locate the owner of the disk
3. try the online path
4. return sane error if online path fails for some reason.

My intention was that the switch between behavioural paths be done in the PL, below even the point of the storage_handlers, as the personality is further from current state and there's no way to prevent state changes.
Doing this in the port layer allows the various Stores to provide switch logic suited to the specific instance, e.g. NFS volume store might _not_ provide a separate online path given locking isn't a consideration.